lib/build.py

```diff
@@ -78,13 +78,28 @@ def main(cli_):
 
    # Infer image name if needed.
    if (cli.tag is None):
-      m = re.search(r"(([^/]+)/)?Dockerfile(\.(.+))?$",
-                    os.path.abspath(cli.file))
-      if (m is not None):
-         if m.group(4):    # extension
-            cli.tag = m.group(4)
-         elif m.group(2):  # containing directory
-            cli.tag = m.group(2)
+      path = os.path.basename(cli.file)
+      if ("." in path):
+         (base, ext_all) = str(path).split(".", maxsplit=1)
+         (base_all, ext_last) = str(path).rsplit(".", maxsplit=1)
+      else:
+         base = None
+         ext_last = None
+      if (base == "Dockerfile"):
+         cli.tag = ext_all
+         ch.VERBOSE("inferring name from Dockerfile extension: %s" % cli.tag)
+      elif (ext_last == "dockerfile"):
+         cli.tag = base_all
+         ch.VERBOSE("inferring name from Dockerfile basename: %s" % cli.tag)
+      elif (os.path.abspath(cli.context) != "/"):
+         cli.tag = os.path.basename(os.path.abspath(cli.context))
+         ch.VERBOSE("inferring name from context directory: %s" % cli.tag)
+      else:
+         assert (os.path.abspath(cli.context) == "/")
+         cli.tag = "root"
+         ch.VERBOSE("inferring name with root context directory: %s" % cli.tag)
+      cli.tag = re.sub(r"[^a-z0-9_.-]", "", cli.tag.lower())
```

Comment:
Currently `:` is stripped so that Dockerfile.foo:v3.5 results in an inferred tag of `foov3.5` is this desired?

Reply:
Good catch. Probably not. The name and tag are colon separated which is why I didn't put it in. Might add a comment saying why the regex doesn't match our documented image reference syntax.

Reply:
Issue #1167.